src/d3a/blockchain/contracts/ink/lib.rs

```diff
@@ -0,0 +1,203 @@
+        fn new(&mut self, initial_supply: Balance) {
+            self.total_supply.set(initial_supply);
+            self.balances.insert(self.env().caller(), initial_supply);
+            self.env().emit_event(
+                Transfer {
+                    from: None,
+                    to: Some(self.env().caller()),
+                    value: initial_supply,
+                });
+        }
+
+        #[ink(message)]
+        fn total_supply(&self) -> Balance {
+            *self.total_supply.get()
+        }
+
+        #[ink(message)]
+        fn balance_of(&self, owner: AccountId) -> Balance {
+            self.balance_of_or_zero(&owner)
+        }
+
+        #[ink(message)]
+        fn approve(&mut self, spender: AccountId, value: Balance) -> bool {
+            let owner = self.env().caller();
+            self.allowances.insert((owner, spender), value);
+            self.env().emit_event(Approval {
+                owner: Some(owner),
+                spender: Some(spender),
+                value: value,
+            });
+            true
+        }
+
+        #[ink(message)]
+        fn allowance(&self, owner: AccountId, spender: AccountId) -> Balance {
+            self.allowance_of_or_zero(&owner, &spender)
+        }
+
+        fn transfer_from(&mut self, from: AccountId, to: AccountId, value: Balance) -> bool {
+            let caller = self.env().caller();
+            let allowance = self.allowance_of_or_zero(&from, &caller);
+            if allowance < value {
+                return false
+            }
+            self.allowances.insert((from, caller), allowance - value);
+            self.transfer_from_to(from, to, value);
+            true
+        }
+
+        #[ink(message)]
+        fn transfer(&mut self, to: AccountId, value: Balance) -> bool {
+            let caller = self.env().caller();
+            self.transfer_from_to(caller, to, value)
+        }
+
+        fn transfer_from_to(&mut self, from: AccountId, to: AccountId, value: Balance) -> bool {
+            let from_balance = self.balance_of_or_zero(&from);
+            let to_balance = self.balance_of_or_zero(&to);
+            if from_balance < value {
+                return false
+            }
+            self.balances.insert(from, from_balance - value);
+            self.balances.insert(to, to_balance + value);
+            self.env()
+                .emit_event(
+                    Transfer {
+                        from: Some(from),
+                        to: Some(to),
+                        value,
+                    });
+            true
+        }
+
+        fn balance_of_or_zero(&self, owner: &AccountId) -> Balance {
+            let balance = self.balances.get(owner).unwrap_or(&0);
+            *balance
+        }
+
+        fn allowance_of_or_zero(&self, owner: &AccountId, spender: &AccountId) -> Balance {
+            let allowance = self.allowances.get(&(*owner, *spender)).unwrap_or(&0);
+            *allowance
+        }
```

Comment:
Same comment(https://github.com/gridsingularity/d3a/pull/883/files#r498653888) goes here too

Reply:
And the same regarding this